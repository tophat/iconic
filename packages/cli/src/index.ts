import * as path from 'path'

import yargs from 'yargs'
import * as fs from 'fs-extra'
import convert from '@react-vector-graphics/core'

function run(src: string, dest: string): void {
    const files = fs.readdirSync(src)
    const filesContents = files.map((file): [string, string] => [
        file,
        fs.readFileSync(path.join(src, file), { encoding: 'utf-8' }),
    ])
    const results = filesContents.map(([file, contents]): object =>
        convert(file, contents),
    )
    for (const result of results) {
        for (const [file, contents] of Object.entries(result)) {
            const filePath = path.join(dest, file)
            fs.outputFileSync(filePath, contents)
        }
    }
}

const argv = yargs
    .usage('Usage: $0 -s [src] -d [dest]')
    .option('src', {
        alias: 's',
        describe: 'source folder',
        type: 'string',
    })
    .option('dest', {
        alias: 'd',
        describe: 'destination folder',
        type: 'string',
    })
    .demandOption(['s', 'd']).argv

const src = argv.s as string
const dest = argv.d as string
run(src, dest)
