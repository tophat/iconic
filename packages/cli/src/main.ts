import yargs from 'yargs'

import { Configuration } from '@react-vector-graphics/types'
import rvgCore from '@react-vector-graphics/core'
import { OPTIONS } from '@react-vector-graphics/plugin-assets'

const mergeCliOptions = (config: Configuration): Configuration => {
    if (!config.options) config.options = {}
    const { argv } = yargs(process.argv)
        .usage('Usage: $0 -p [pattern] -o [output]')
        .option('pattern', {
            alias: 'p',
            default: config.options[OPTIONS.GLOB_PATTERN],
            describe: 'SVG files glob pattern',
            type: 'string',
        })
        .option('output', {
            alias: 'o',
            default: config.options[OPTIONS.OUTPUT_PATH],
            describe: 'Destination folder',
            type: 'string',
        })
        .option('ext', {
            alias: 'e',
            default: config.options[OPTIONS.FILE_EXT],
            describe: 'Component file extension',
            type: 'string',
        })
        .demandOption(['pattern', 'output'])

    config.options[OPTIONS.GLOB_PATTERN] = argv.pattern
    config.options[OPTIONS.OUTPUT_PATH] = argv.output
    config.options[OPTIONS.FILE_EXT] = argv.ext
    return config
}

export const run = async (config: Configuration): Promise<void> => {
    const assetPlugin = '@react-vector-graphics/plugin-assets'
    if (!config.plugins?.length) {
        config.plugins = [assetPlugin, '@svgr/plugin-jsx', assetPlugin]
    }
    await rvgCore({ config: mergeCliOptions(config) })
}
