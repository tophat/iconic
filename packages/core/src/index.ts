export { getLogger } from './logging'
export { run as default } from './main'
