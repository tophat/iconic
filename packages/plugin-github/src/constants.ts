const ns = (s: TemplateStringsArray): string => `github/${s}`

export const OPTIONS = {
    API: ns`api`,
    BASE: ns`base`,
    COMMIT_CREATE_PATTERN: ns`commitCreate`,
    COMMIT_DELETE_PATTERN: ns`commitDelete`,
    COMMIT_UPDATE_PATTERN: ns`commitUpdate`,
    FILE_EXT: ns`fileExt`,
    FOLDER_PATH: ns`folderPath`,
    GLOB_PATTERN: ns`globPattern`,
    HEAD: ns`head`,
    NAME_SCHEME: ns`nameScheme`,
    OUTPUT_PATH: ns`outputPath`,
    OWNER: ns`owner`,
    REPO: ns`repo`,
}

export const STATE = {
    COMPONENT_FILES: ns`componentFiles`,
    COMPONENT_NAME: 'componentName', // @svgr/plugin-jsx
    COMPONENT_NAME_OLD: ns`componentNameOld`,
    DIFF_TYPE: ns`diffType`,
    FILE_PATH: 'filePath', // @svgr/plugin-jsx
}

export const STATUSES = {
    ADDED: 'added',
    MODIFIED: 'modified',
    REMOVED: 'removed',
    RENAMED: 'renamed',
}

export const COMMIT_MESSAGE_PLACEHOLDER = '${message}'

export const COMMIT_MESSAGE_PATTERNS = {
    CREATE: `feat: ${COMMIT_MESSAGE_PLACEHOLDER}`,
    DELETE: `refactor: ${COMMIT_MESSAGE_PLACEHOLDER}\nBREAKING CHANGE: ${COMMIT_MESSAGE_PLACEHOLDER}`,
    UPDATE: `fix: ${COMMIT_MESSAGE_PLACEHOLDER}`,
}
