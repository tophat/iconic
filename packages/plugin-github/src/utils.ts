import { Logger } from '@react-vector-graphics/types'

export const fromBase64 = (data: string): string =>
    Buffer.from(data, 'base64').toString()

export const toBase64 = (data: string): string =>
    Buffer.from(data).toString('base64')

export const eagerPromises = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    promises: Promise<any>[],
    logger: Logger = console,
): // eslint-disable-next-line @typescript-eslint/no-explicit-any
Promise<any> => Promise.all(promises.map(p => p.catch(logger.error)))
