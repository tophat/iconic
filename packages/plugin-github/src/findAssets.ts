import * as path from 'path'

import minimatch from 'minimatch'
import { Octokit } from '@octokit/rest'

import { NamingScheme, PluginParams, State } from '@react-vector-graphics/types'
import { pathToName } from '@react-vector-graphics/utils'

import { STATE } from './constants'
import { fromBase64, normaliseGlob } from './utils'

const findAssets = async ({
    folderPath = '',
    github: { api: githubApi, ...githubParams },
    ...params
}: {
    folderPath: string
    github: {
        api: Octokit
        base: string
        head: string
        owner: string
        repo: string
    }
    globPattern: string
    nameScheme: NamingScheme
    state: State
}): Promise<PluginParams[]> => {
    const compareCommitsResult = await githubApi.repos.compareCommits(
        githubParams,
    )
    const svgFiles = compareCommitsResult.data.files.filter(file => {
        const isInFolder = file.filename.startsWith(folderPath)
        if (!isInFolder) return false
        const relPath = path.relative(folderPath, file.filename)
        return minimatch(relPath, normaliseGlob(params.globPattern))
    })
    const pluginParams = svgFiles.map(
        async (file): Promise<PluginParams> => {
            const filePath = path.relative(folderPath, file.filename)
            const { data } = await githubApi.repos.getContents({
                ...githubParams,
                path: file.filename,
                ref: githubParams.head,
            })
            if (Array.isArray(data) || !data.content) {
                throw new Error(`Could not get contents for ${file.filename}`)
            }
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            const { previous_filename: previousFilename } = file
            return {
                code: fromBase64(data.content.toString()),
                state: Object.assign(
                    {
                        [STATE.COMPONENT_NAME]: pathToName(
                            filePath,
                            params.nameScheme,
                        ),
                        [STATE.COMPONENT_NAME_OLD]:
                            previousFilename &&
                            pathToName(
                                path.relative(folderPath, previousFilename),
                                params.nameScheme,
                            ),
                        [STATE.DIFF_TYPE]: file.status,
                        [STATE.FILE_PATH]: filePath,
                    },
                    params.state,
                ),
            }
        },
    )
    return Promise.all(pluginParams)
}

export default findAssets
