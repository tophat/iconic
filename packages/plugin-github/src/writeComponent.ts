import * as path from 'path'
import { promises } from 'dns'

import { Octokit } from '@octokit/rest'

import { Logger } from '@react-vector-graphics/types'

import {
    COMMIT_MESSAGE_PATTERNS,
    COMMIT_MESSAGE_PLACEHOLDER,
    OPTIONS,
    STATE,
    STATUSES,
} from './constants'
import { eagerPromises, replaceAll, toBase64 } from './utils'

const removeIconFiles = async (
    githubApi: Octokit,
    githubParams: { head: string; owner: string; repo: string },
    componentName: string,
    componentPath: string,
    commitMessagePattern: string = COMMIT_MESSAGE_PATTERNS.DELETE,
): Promise<void> => {
    const deleteMessage = `remove ${componentName}`
    const message = replaceAll(
        commitMessagePattern,
        COMMIT_MESSAGE_PLACEHOLDER,
        deleteMessage,
    )
    const { data: results } = await githubApi.repos.getContents({
        ...githubParams,
        path: componentPath,
        ref: githubParams.head,
    })
    await eagerPromises(
        (Array.isArray(results) ? results : [results]).map(({ path, sha }) =>
            githubApi.repos.deleteFile({
                ...githubParams,
                message,
                path,
                sha,
            }),
        ),
    )
}

const addOrModifyIconFile = async (
    githubApi: Octokit,
    githubParams: { head: string; owner: string; repo: string },
    componentName: string,
    fileName: string,
    filePath: string,
    fileContents: string,
    commitMessagePatternCreate: string = COMMIT_MESSAGE_PATTERNS.CREATE,
    commitMessagePatternUpdate: string = COMMIT_MESSAGE_PATTERNS.UPDATE,
): Promise<void> => {
    let fileSha
    try {
        const { data } = await githubApi.repos.getContents({
            ...githubParams,
            path: filePath,
            ref: githubParams.head,
        })
        if (Array.isArray(data)) return
        fileSha = data.sha
    } catch (e) {
        // assume file does not exist and do nothing
    }
    const message = fileSha
        ? replaceAll(
              commitMessagePatternUpdate,
              COMMIT_MESSAGE_PLACEHOLDER,
              `modify ${componentName} ${fileName}`,
          )
        : replaceAll(
              commitMessagePatternCreate,
              COMMIT_MESSAGE_PLACEHOLDER,
              `add ${componentName} ${fileName}`,
          )
    await githubApi.repos.createOrUpdateFile({
        ...githubParams,
        branch: githubParams.head,
        content: toBase64(fileContents),
        message,
        path: filePath,
        sha: fileSha,
    })
}

const writeComponent = async ({
    github: { api: githubApi, ...githubParams },
    ...params
}: {
    assetFile: string
    code: string
    commitMessagePatterns?: {
        create?: string
        delete?: string
        update?: string
    }
    componentName?: string
    componentNameOld?: string
    componentFiles: { [fileName: string]: string }
    diffType: string
    fileExt?: string
    folderPath: string
    github: {
        api: Octokit
        base: string
        head: string
        owner: string
        repo: string
    }
    logger?: Logger
    outputPath?: string
}): Promise<void> => {
    if (!params.componentName) {
        return params.logger?.warn(
            `No '${STATE.COMPONENT_NAME}' provided for '${params.assetFile}'.`,
        )
    }
    if (!params.outputPath) {
        return params.logger?.warn(`No '${OPTIONS.OUTPUT_PATH}' provided.`)
    }
    if (!params.fileExt) {
        params.logger?.warn(`No '${OPTIONS.FILE_EXT}' provided.`)
    }
    // gather files
    const componentFiles = Object.entries(params.componentFiles)
    const singleFile = componentFiles.length === 0
    const pathToFolder = path.join(
        params.folderPath,
        params.outputPath,
        singleFile ? '' : params.componentName,
    )
    const componentFileName =
        (singleFile ? params.componentName : 'index') +
        (params.fileExt ? `.${params.fileExt}` : '')
    componentFiles.push([componentFileName, params.code])
    const componentFilePath = path.join(pathToFolder, componentFileName)
    // commit file changes
    const pendingPromises = []
    if (params.diffType === STATUSES.REMOVED) {
        pendingPromises.push(
            removeIconFiles(
                githubApi,
                githubParams,
                params.componentName,
                singleFile ? componentFilePath : pathToFolder,
                params.commitMessagePatterns?.delete,
            ),
        )
    } else {
        // added, modified or renamed
        for (const [fileName, fileContents] of componentFiles) {
            const filePath = path.join(pathToFolder, fileName)
            pendingPromises.push(
                addOrModifyIconFile(
                    githubApi,
                    githubParams,
                    params.componentName,
                    fileName,
                    filePath,
                    fileContents,
                    params.commitMessagePatterns?.create,
                    params.commitMessagePatterns?.update,
                ),
            )
        }
        if (params.diffType === STATUSES.RENAMED && params.componentNameOld) {
            const oldPathToFolder = path.join(
                params.folderPath,
                params.outputPath,
                singleFile ? '' : params.componentNameOld,
            )
            const oldPathToFile = path.join(
                oldPathToFolder,
                singleFile ? params.componentNameOld : 'index',
            )
            const oldComponentFilePath = params.fileExt
                ? `${oldPathToFile}.${params.fileExt}`
                : oldPathToFile
            pendingPromises.push(
                removeIconFiles(
                    githubApi,
                    githubParams,
                    params.componentNameOld,
                    singleFile ? oldComponentFilePath : oldPathToFolder,
                    params.commitMessagePatterns?.delete,
                ),
            )
        }
    }
    await eagerPromises(pendingPromises)
}

export default writeComponent
