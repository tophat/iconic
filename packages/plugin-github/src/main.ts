import { Plugin } from '@react-vector-graphics/types'

import { OPTIONS, STATE } from './constants'
import findAssets from './findAssets'
import writeComponent from './writeComponent'

export const run: Plugin = async (code, config, state, logger) => {
    const github = {
        api: config.options[OPTIONS.API],
        base: config.options[OPTIONS.BASE],
        head: config.options[OPTIONS.HEAD],
        owner: config.options[OPTIONS.OWNER],
        repo: config.options[OPTIONS.REPO],
    }
    const sharedArgs = {
        folderPath: config.options[OPTIONS.FOLDER_PATH],
        github,
    }
    if (code) {
        await writeComponent({
            ...sharedArgs,
            assetFile: state[STATE.FILE_PATH],
            code,
            commitMessagePatterns: {
                create: config.options[OPTIONS.COMMIT_CREATE_PATTERN],
                delete: config.options[OPTIONS.COMMIT_DELETE_PATTERN],
                update: config.options[OPTIONS.COMMIT_UPDATE_PATTERN],
            },
            componentFiles: state[STATE.COMPONENT_FILES] ?? {},
            componentName: state[STATE.COMPONENT_NAME],
            componentNameOld: state[STATE.COMPONENT_NAME_OLD],
            diffType: state[STATE.DIFF_TYPE],
            fileExt: config.options[OPTIONS.FILE_EXT],
            logger,
            outputPath: config.options[OPTIONS.OUTPUT_PATH],
        })
        return { code, state }
    } else {
        return findAssets({
            ...sharedArgs,
            globPattern: config.options[OPTIONS.GLOB_PATTERN],
            nameScheme: config.options[OPTIONS.NAME_SCHEME],
            state,
        })
    }
}
