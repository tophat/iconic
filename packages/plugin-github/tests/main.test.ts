import * as findAssets from '../src/findAssets'
import * as writeComponent from '../src/writeComponent'
import { OPTIONS, STATE } from '../src/constants'
import { run } from '../src/main'

import { mockComponent, mockOptions, mockSVG, mockState } from './mocks'

describe('main', () => {
    beforeAll(() => {
        jest.spyOn(findAssets, 'default').mockResolvedValue([
            {
                code: mockSVG,
                state: {
                    ...mockState,
                    [STATE.COMPONENT_FILES]: {},
                },
            },
        ])
        jest.spyOn(writeComponent, 'default').mockResolvedValue()
    })
    afterEach(jest.resetAllMocks)
    afterAll(jest.restoreAllMocks)

    it('finds assets if no code passed', async () => {
        const results = run(
            undefined,
            { options: mockOptions, plugins: [] },
            {},
            console,
        )
        await expect(results).resolves.toMatchSnapshot()
    })

    it('writes component files when code passed', async () => {
        const results = run(
            mockComponent,
            { options: mockOptions, plugins: [] },
            mockState,
            console,
        )
        await expect(results).resolves.toMatchSnapshot()
        expect(writeComponent.default).toHaveBeenCalledWith({
            assetFile: mockState[STATE.FILE_PATH],
            code: mockComponent,
            commitMessagePatterns: {
                create: mockOptions[OPTIONS.COMMIT_CREATE_PATTERN],
                delete: mockOptions[OPTIONS.COMMIT_DELETE_PATTERN],
                update: mockOptions[OPTIONS.COMMIT_UPDATE_PATTERN],
            },
            componentFiles: mockState[STATE.COMPONENT_FILES],
            componentName: mockState[STATE.COMPONENT_NAME],
            componentNameOld: mockState[STATE.COMPONENT_NAME_OLD],
            diffType: mockState[STATE.DIFF_TYPE],
            fileExt: mockOptions[OPTIONS.FILE_EXT],
            folderPath: mockOptions[OPTIONS.FOLDER_PATH],
            github: {
                api: mockOptions[OPTIONS.API],
                base: mockOptions[OPTIONS.BASE],
                head: mockOptions[OPTIONS.HEAD],
                owner: mockOptions[OPTIONS.OWNER],
                repo: mockOptions[OPTIONS.REPO],
            },
            logger: console,
            outputPath: mockOptions[OPTIONS.OUTPUT_PATH],
        })
    })
})
