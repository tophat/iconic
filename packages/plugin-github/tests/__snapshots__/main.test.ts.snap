// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main finds assets if no code passed 1`] = `
Array [
  Object {
    "code": "<svg>mock</svg>",
    "state": Object {
      "componentName": "mockIcon",
      "filePath": "assets/mock.icon.svg",
      "github/componentFiles": Object {},
      "github/diffType": "added",
    },
  },
]
`;

exports[`main writes component files when code passed 1`] = `
Object {
  "code": "
import React from 'react'
export default function mockIcon() {
    return <svg>mock updated</svg>
}
",
  "state": Object {
    "componentName": "mockIcon",
    "filePath": "assets/mock.icon.svg",
    "github/componentFiles": Object {
      "README.md": "## mockIcon

mock usage notes",
    },
    "github/diffType": "added",
  },
}
`;
